metals/src/main/scala/scala/meta/internal/decorations/SyntheticHoverProvider.scala

```diff
@@ -0,0 +1,279 @@
+package scala.meta.internal.decorations
+
+import java.nio.charset.Charset
+
+import scala.util.Try
+
+import scala.meta.internal.metals.Buffers
+import scala.meta.internal.metals.ClientCommands
+import scala.meta.internal.metals.ClientConfiguration
+import scala.meta.internal.metals.CommandHTMLFormat
+import scala.meta.internal.metals.HoverExtParams
+import scala.meta.internal.metals.MetalsEnrichments._
+import scala.meta.internal.metals.ServerCommands
+import scala.meta.internal.metals.UserConfiguration
+import scala.meta.internal.metap.PrinterSymtab
+import scala.meta.internal.mtags.Md5Fingerprints
+import scala.meta.internal.mtags.Semanticdbs
+import scala.meta.internal.parsing.Trees
+import scala.meta.internal.pc.HoverMarkup
+import scala.meta.internal.semanticdb.Scala._
+import scala.meta.internal.semanticdb.TextDocument
+import scala.meta.internal.{semanticdb => s}
+import scala.meta.io.AbsolutePath
+
+import org.eclipse.{lsp4j => l}
+
+final class SyntheticHoverProvider(
+    workspace: AbsolutePath,
+    semanticdbs: Semanticdbs,
+    buffer: Buffers,
+    fingerprints: Md5Fingerprints,
+    charset: Charset,
+    clientConfig: ClientConfiguration,
+    userConfig: () => UserConfiguration,
+    trees: Trees,
+) {
+
+  @volatile private var document: Option[TextDocument] =
+    Option.empty[TextDocument]
+
+  def addSyntheticsHover(
+      params: HoverExtParams,
+      pcHover: Option[l.Hover],
+  ): Option[l.Hover] =
+    if (userConfig().areSyntheticsEnabled()) {
+      val path = params.textDocument.getUri().toAbsolutePath
+      val position = params.getPosition
+      val line = position.getLine()
+      val isInlineDecorationProvider = clientConfig.isInlineDecorationProvider()
+      val newHover = currentDocument(path) match {
```

Comment:
The hover is coming from Semanticdb, but the decoration itself from presentation compiler, will that not be a problem?

Reply:
Decorations from PC are at least as good as those from semantic db, so there shouldn’t be a case that there is a decoration from semanticdb and not from PC. It's possible that there will no hover on some decorations (ie. `given` arguments), because they are not in semanticdb synthetics. We could try to move hover to PC, but since we want to use inlay hints either way I don't think it's worth it?